File: src/javascript/path-traversal.js
// CWE-22: Path Traversal.

const express = require('express');
const fs = require('fs');
const path = require('path');

const app = express();

app.get('/download', (req, res) => {
  const file = req.query.file;
  const data = fs.readFileSync('/var/app/uploads/' + file);
  res.send(data);
});

app.get('/template', (req, res) => {
  const name = req.query.name;
  const tpl = fs.readFileSync(path.join('/var/app/templates', name), 'utf8');
  res.type('text/html').send(tpl);
});

app.get('/static/*', (req, res) => {
  const requested = req.params[0];
  const full = '/var/app/public/' + requested;
  res.sendFile(full);
});

app.listen(3003);

File: src/python/path_traversal.py
"""CWE-22: Path Traversal."""

import os
from flask import Flask, request, send_file

app = Flask(__name__)
UPLOADS = "/var/app/uploads"


@app.route("/download")
def download():
    name = request.args.get("file", "")
    path = os.path.join(UPLOADS, name)
    with open(path, "rb") as f:
        return f.read()


@app.route("/template")
def template():
    name = request.args.get("name", "")
    return send_file("/var/app/templates/" + name)


@app.route("/log")
def log():
    log_id = request.args.get("id", "")
    full = "/var/log/app/" + log_id + ".log"
    return open(full).read()


if __name__ == "__main__":
    app.run(port=4003)
